hydra_base/lib/users.py

```diff
@@ -174,6 +176,54 @@ def delete_user(deleted_user_id,**kwargs):
 
     return 'OK'
 
+def get_failed_login_count(username, **kwargs):
+    try:
+        user_i = db.DBSession.query(User).filter(User.username == username).one()
+    except NoResultFound:
+        """ Non-existent user should not raise """
+        return -1
+
+    return user_i.failed_logins
+
+
+def get_max_login_attempts(*args, **kwargs):
+    max_login_attempts = int(config.get("security", "max_login_attempts"))
+
+    return max_login_attempts if max_login_attempts is not None else 0
+
+
+def get_remaining_login_attempts(username, **kwargs):
+    max_login_attempts = get_max_login_attempts()
+    failed_attempts = get_failed_login_count(username)
+
+    return max_login_attempts - failed_attempts
+
+
+def inc_failed_login_attempts(username, **kwargs):
+    try:
+        user_i = db.DBSession.query(User).filter(User.username == username).one()
+    except NoResultFound:
+        raise HydraError(username)
+
+    if user_i.failed_logins is not None:
+        user_i.failed_logins = user_i.failed_logins + 1
+    else:
+        user_i.failed_logins = 1
+
+    db.DBSession.flush()
+    transaction.commit()
```

Comment:
You should use the commit_transaction() function in hydra base DB module instead.